### PR TITLE
En 4654 stop sync and consensus when not connected

### PR DESCRIPTION
**Pull request by iulianpascalau**
Sync and consensus halt if the number of connected peers is lower than a minimum number.

### CODE REVIEW

File: p2p/libp2p/libp2pConnectionMonitor.go [hunk 1]
```diff
@@ -8,23 +8,21 @@ import (
...
 }
 
-func newLibp2pConnectionMonitor(reconnecter p2p.Reconnecter) *libp2pConnectionMonitor {
+func newLibp2pConnectionMonitor(reconnecter p2p.Reconnecter, thresholdMinConnectedPeers int) *libp2pConnectionMonitor {
 	cm := &libp2pConnectionMonitor{
-		reconnecter:   reconnecter,
-		chDoReconnect: make(chan struct{}, 0),
+		reconnecter:                reconnecter,
+		chDoReconnect:              make(chan struct{}, 0),
+		thresholdMinConnectedPeers: thresholdMinConnectedPeers,
```
Comment (bogdan-rosianu) [inline p2p/libp2p/libp2pConnectionMonitor.go]: You can add a check for this new field (not negative, stuff like that) and return an error if the value is wrong. Also, add a test for this case

File: p2p/libp2p/netMessenger.go [hunk 2]
```diff
@@ -509,10 +512,25 @@ func (netMes *networkMessenger) directMessageHandler(message p2p.MessageP2P) err
...
+// IsConnectedToTheNetwork returns true if the current node is connected to the network
+func (netMes *networkMessenger) IsConnectedToTheNetwork() bool {
+	netw := netMes.ctxProvider.connHost.Network()
+	return netMes.connMonitor.isConnectedToTheNetwork(netw)
+}
+
+// SetThresholdMinConnectedPeers sets the minimum connected peers before triggering a new reconnection
+func (netMes *networkMessenger) SetThresholdMinConnectedPeers(minConnectedPeers int) {
+	netw := netMes.ctxProvider.connHost.Network()
+	netMes.connMonitor.thresholdMinConnectedPeers = minConnectedPeers
```
Comment (bogdan-rosianu) [inline p2p/libp2p/netMessenger.go]: add a check for _minConnectedPeers_

File: process/sync/metablock.go [hunk 3]
```diff
@@ -369,6 +372,11 @@ func (boot *MetaBootstrap) StopSync() {
 func (boot *MetaBootstrap) syncBlocks() {
 	for {
 		time.Sleep(sleepTime)
+
+		if !boot.networkWatcher.IsConnectedToTheNetwork() {
+			continue
```
Comment (bogdan-rosianu) [inline process/sync/metablock.go]: add a test for this. All the tests in the test file return true for `IsConnectedToTheNetwork`

File: process/sync/shardblock.go [hunk 4]
```diff
@@ -599,6 +602,11 @@ func (boot *ShardBootstrap) StopSync() {
 func (boot *ShardBootstrap) syncBlocks() {
 	for {
 		time.Sleep(sleepTime)
+
+		if !boot.networkWatcher.IsConnectedToTheNetwork() {
+			continue
```
Comment (bogdan-rosianu) [inline process/sync/shardblock.go]: also here

File: process/sync/baseSync.go [hunk 5]
```diff
@@ -449,10 +449,16 @@ func (boot *baseBootstrap) waitForHeaderHash() error {
 	}
 }
 
-// ShouldSync method returns the synch state of the node. If it returns 'true', this means that the node
+// ShouldSync method returns the sync state of the node. If it returns 'true', this means that the node
 // is not synchronized yet and it has to continue the bootstrapping mechanism, otherwise the node is already
-// synched and it can participate to the consensus, if it is in the jobDone group of this rounder
+// synced and it can participate to the consensus, if it is in the jobDone group of this rounder
+// should note that when the node is not connected to the network, ShouldSync returns true but the SyncBlock
```
Comment (bogdan-rosianu) [inline process/sync/baseSync.go]: This is a new idea, so you can capitalize it for readability reasons

File: integrationTests/testInitializer.go [hunk 6]
```diff
@@ -328,7 +328,7 @@ func CreateAccount(accnts state.AccountsAdapter, nonce uint64, balance *big.Int)
 // MakeDisplayTable will output a string containing counters for received transactions, headers, miniblocks and
 // meta headers for all provided test nodes
 func MakeDisplayTable(nodes []*TestProcessorNode) string {
-	header := []string{"pk", "shard ID", "txs", "miniblocks", "headers", "metachain headers"}
+	header := []string{"pk", "shard ID", "txs", "miniblocks", "headers", "metachain headers", "conns"}
```
Comment (sasurobert) [inline integrationTests/testInitializer.go]: connections

File: p2p/libp2p/libp2pConnectionMonitor.go (hunk 1 above)
Comment (sasurobert) [reply p2p/libp2p/libp2pConnectionMonitor.go]: +1

File: p2p/libp2p/libp2pConnectionMonitor.go [hunk 7]
```diff
@@ -69,3 +71,8 @@ func (lcm *libp2pConnectionMonitor) doReconnection() {
 		time.Sleep(DurationBetweenReconnectAttempts)
 	}
 }
+
+func (lcm *libp2pConnectionMonitor) isConnectedToTheNetwork(netw network.Network) bool {
+	numConnections := netw.Conns()
+	return len(numConnections) >= lcm.thresholdMinConnectedPeers
```
Comment (sasurobert) [inline p2p/libp2p/libp2pConnectionMonitor.go]: you could make it a one liner.

File: p2p/libp2p/netMessenger.go (hunk 2 above)
Comment (sasurobert) [reply p2p/libp2p/netMessenger.go]: +1

File: p2p/p2p.go [hunk 8]
```diff
@@ -155,6 +155,10 @@ type Messenger interface {
 	// peer, but reuses a connection and a stream if possible.
 	SendToConnectedPeer(topic string, buff []byte, peerID PeerID) error
 
+	IsConnectedToTheNetwork() bool
```
Comment (sasurobert) [inline p2p/p2p.go]: IsConnected would be enough

File: process/interface.go [hunk 9]
```diff
@@ -478,3 +478,10 @@ type MiniBlocksCompacter interface {
 	Expand(block.MiniBlockSlice, map[string]data.TransactionHandler) (block.MiniBlockSlice, error)
 	IsInterfaceNil() bool
 }
+
+// NetworkConnectionWatcher defines a watchdog functionality used to specify if the current node
+// is still connected to the rest of the network
+type NetworkConnectionWatcher interface {
+	IsConnectedToTheNetwork() bool
```
Comment (sasurobert) [inline process/interface.go]: IsConnected ?

File: p2p/libp2p/libp2pConnectionMonitor.go (hunk 1 above)
Comment (iulianpascalau) [reply p2p/libp2p/libp2pConnectionMonitor.go]: Done

File: p2p/libp2p/netMessenger.go (hunk 2 above)
Comment (iulianpascalau) [reply p2p/libp2p/netMessenger.go]: Done

File: process/sync/metablock.go (hunk 3 above)
Comment (iulianpascalau) [reply process/sync/metablock.go]: Will be done in this PR at a latter time. It depends on a pending refactoring.

File: process/sync/baseSync.go (hunk 5 above)
Comment (iulianpascalau) [reply process/sync/baseSync.go]: Done

File: integrationTests/testInitializer.go (hunk 6 above)
Comment (iulianpascalau) [reply integrationTests/testInitializer.go]: done

File: p2p/libp2p/libp2pConnectionMonitor.go (hunk 7 above)
Comment (iulianpascalau) [reply p2p/libp2p/libp2pConnectionMonitor.go]: done

File: p2p/p2p.go (hunk 8 above)
Comment (iulianpascalau) [reply p2p/p2p.go]: We already have the function `IsConnected(peerID PeerID) bool`

File: process/interface.go (hunk 9 above)
Comment (iulianpascalau) [reply process/interface.go]: We already have the function `IsConnected(peerID PeerID) bool`

**Pull request merged**